File: 卡码/代码随想录/1/11-图论/Java/src/com/kurumi/bfs/LeetCode_0695.java
//695. 岛屿的最大面积
package com.kurumi.bfs;

import java.util.ArrayDeque;
import java.util.Deque;

public class LeetCode_0695 {

}
class Solution {
	int[][] dir=new int[][] {{0,1},{1,0},{0,-1},{-1,0}};
	int cnt;
	public void bfs(int[][] grid,int x,int y) {
		grid[x][y]=0;
		Deque<int[]> q=new ArrayDeque<>();
		q.offer(new int[] {x,y});
		while(!q.isEmpty()) {
			int[] v=q.poll();
			for(int i=0;i<4;i++) {
				int nextX=v[0]+dir[i][0];
				int nextY=v[1]+dir[i][1];
				if(nextX<0||nextX>grid.length-1||nextY<0||nextY>grid[0].length-1) continue;
				if(grid[nextX][nextY]==1) {
					cnt++;
					grid[nextX][nextY]=0;
					q.offer(new int[] {nextX,nextY});
				}
			}
		}
	}
  public int maxAreaOfIsland(int[][] grid) {
  	int n=grid.length;
  	int m=grid[0].length;
  	int res=0;
  	for(int i=0;i<n;i++) {
  		for(int j=0;j<m;j++) {
  			if(grid[i][j]==1) {
  				cnt=1;
  				bfs(grid,i,j);
  				res=Math.max(res,cnt);
  			}
  		}
  	}
  	return res;
  }
}

File: 卡码/代码随想录/1/11-图论/Python/bfs/LeetCode_0695.py
#695. 岛屿的最大面积
from typing import *
from collections import deque

class Solution:
	def __init__(self):
		self.dir:List[tuple[int,int]]=[(0,1),(1,0),(0,-1),(-1,0)]
		self.cnt:int=1
	def bfs(self,grid:List[List[int]],x:int,y:int):
		grid[x][y]=0
		q:deque[tuple[int,int]]=deque()
		q.append((x,y))
		while len(q)>0:
			v:tuple[int,int]=q.pop()
			for i in range(4):
				nextX:int=v[0]+self.dir[i][0]
				nextY:int=v[1]+self.dir[i][1]
				if nextX<0 or nextX>len(grid)-1 or nextY<0 or nextY>len(grid[0])-1:
					continue
				if grid[nextX][nextY]==1:
					self.cnt+=1
					grid[nextX][nextY]=0
					q.append((nextX,nextY))
	def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
		n,m=len(grid),len(grid[0])
		res:int=0
		for i in range(n):
			for j in range(m):
				if grid[i][j]==1:
					self.cnt=1
					self.bfs(grid,i,j)
					res=max(res,self.cnt)
		return res